crates/apollo-compiler/src/execution/engine.rs

```diff
@@ -0,0 +1,327 @@
+use crate::execution::JsonMap;
+use crate::execution::JsonValue;
+use crate::execution::ResponseDataPathElement;
+use crate::node::NodeLocation;
+use crate::schema::ExtendedType;
+use crate::schema::FieldDefinition;
+use crate::schema::ObjectType;
+use crate::schema::Type;
+use crate::validation::SuspectedValidationBug;
+use crate::validation::Valid;
+use crate::ExecutableDocument;
+use crate::Schema;
+use crate::SourceMap;
+use indexmap::IndexMap;
+use std::collections::HashSet;
+
+/// <https://spec.graphql.org/October2021/#sec-Normal-and-Serial-Execution>
+#[derive(Debug, Copy, Clone)]
+pub(crate) enum ExecutionMode {
+    /// Allowed to resolve fields in any order, including in parellel
+    Normal,
+    /// Top-level fields of a mutation operation must be executed in order
+    #[allow(unused)]
+    Sequential,
+}
+
+/// Return in `Err` when a field error occurred at some non-nullable place
+///
+/// <https://spec.graphql.org/October2021/#sec-Handling-Field-Errors>
+pub(crate) struct PropagateNull;
+
+/// Linked-list version of `Vec<PathElement>`, taking advantage of the call stack
+pub(crate) type LinkedPath<'a> = Option<&'a LinkedPathElement<'a>>;
+
+pub(crate) struct LinkedPathElement<'a> {
+    pub(crate) element: ResponseDataPathElement,
+    pub(crate) next: LinkedPath<'a>,
+}
+
+/// <https://spec.graphql.org/October2021/#ExecuteSelectionSet()>
+#[allow(clippy::too_many_arguments)] // yes it’s not a nice API but it’s internal
+pub(crate) fn execute_selection_set<'a>(
+    schema: &Valid<Schema>,
+    document: &'a Valid<ExecutableDocument>,
+    variable_values: &Valid<JsonMap>,
+    errors: &mut Vec<GraphQLError>,
+    path: LinkedPath<'_>,
+    mode: ExecutionMode,
+    object_type_name: &str,
+    object_type: &ObjectType,
+    object_value: &ObjectValue<'_>,
+    selections: impl IntoIterator<Item = &'a Selection>,
+) -> Result<JsonMap, PropagateNull> {
+    let mut grouped_field_set = IndexMap::new();
+    collect_fields(
+        schema,
+        document,
+        variable_values,
+        object_type_name,
+        object_type,
+        selections,
+        &mut HashSet::new(),
+        &mut grouped_field_set,
+    );
+
+    match mode {
+        ExecutionMode::Normal => {}
+        ExecutionMode::Sequential => {
+            // If we want parallelism, use `futures::future::join_all` (async)
+            // or Rayon’s `par_iter` (sync) here.
+        }
+    }
+
+    let mut response_map = JsonMap::with_capacity(grouped_field_set.len());
+    for (&response_key, fields) in &grouped_field_set {
+        // Indexing should not panic: `collect_fields` only creates a `Vec` to push to it
+        let field_name = &fields[0].name;
+        let Ok(field_def) = schema.type_field(object_type_name, field_name) else {
+            // TODO: Return a `validation_bug`` field error here?
+            // The spec specifically has a “If fieldType is defined” condition,
+            // but it being undefined would make the request invalid, right?
```

Comment:
I don’t think so. This is about [field errors](https://spec.graphql.org/October2021/#sec-Handling-Field-Errors) that happen during execution. An undefined field makes an operation invalid, so execution shouldn’t even start.

This comment is because I was surprised to see the spec use "if" there, whereas it uses "assert" in other places ([example](https://spec.graphql.org/October2021/#sel-IANNFHCDABABGqD)) for things expected of valid documents.

Reply:
gotcha gotcha